aws/resources/ec2.go

```diff
@@ -140,722 +130,34 @@ func (ei *EC2Instances) nukeAll(instanceIds []*string) error {
 	logging.Debugf("Terminating all EC2 instances in region %s", ei.Region)
 
 	params := &ec2.TerminateInstancesInput{
-		InstanceIds: instanceIds,
+		InstanceIds: aws.ToStringSlice(instanceIds),
 	}
 
-	_, err = ei.Client.TerminateInstancesWithContext(ei.Context, params)
+	_, err = ei.Client.TerminateInstances(ei.Context, params)
 	if err != nil {
 		logging.Debugf("[Failed] %s", err)
 		return errors.WithStackTrace(err)
 	}
 
-	err = ei.Client.WaitUntilInstanceTerminatedWithContext(ei.Context, &ec2.DescribeInstancesInput{
-		Filters: []*ec2.Filter{
+	waiter := ec2.NewInstanceTerminatedWaiter(ei.Client)
+	err = waiter.Wait(ei.Context, &ec2.DescribeInstancesInput{
+		Filters: []types.Filter{
 			{
-				Name:   awsgo.String("instance-id"),
-				Values: instanceIds,
+				Name:   aws.String("instance-id"),
+				Values: aws.ToStringSlice(instanceIds),
 			},
 		},
-	})
-	for _, instanceID := range instanceIds {
-		logging.Debugf("Terminated EC2 Instance: %s", *instanceID)
-	}
+	}, 15*time.Minute)
 
 	if err != nil {
 		logging.Debugf("[Failed] %s", err)
 		return errors.WithStackTrace(err)
 	}
 
-	logging.Debugf("[OK] %d instance(s) terminated in %s", len(instanceIds), ei.Region)
-	return nil
-}
-
-func GetEc2ServiceClient(region string) ec2iface.EC2API {
```

Comment:
To be honest, I was not sure why there is everything below this line when there is already a separate `ec2_vpc.go` file?

Reply:
I actually think this is some old or dead code,
That function is called from `GetDefaultSecurityGroups` and `NewVpcPerRegion`, but I am unable to locate calls to them. This likely needs to be cleaned up or removed, or the signatures should be fixed to allow the code to build.

Reply:
@wakeful Yes, that's what I figured. In Goland I see something like the following for many of those functions:

<img width="666" alt="Screenshot 2024-11-28 at 06 25 25" src="https://github.com/user-attachments/assets/a63d3a32-0ac9-413c-804e-ac6f50a73311">

These functions are greyed out:

- NewVpcPerRegion
- GetDefaultVpcs
- NukeVpcs
- GetDefaultSecurityGroups
- NukeDefaultSecurityGroupRules

Then all of the `nuke*` functions are called by the main `(v Vpc) nuke` function in that file. These `nuke*` functions also exist in other places. For example:

- `nukeInternetGateways` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L325
- `nukeEgressOnlyGateways` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L357
- `nukeEndpoints` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L393
- `nukeNetworkInterfaces` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L433
- `nukeSubnets` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L474
- `nukeRouteTables` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L539
- `nukeNacls` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L600
- `nukeSecurityGroups` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L656
- `dissociateDhcpOptions` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L738
- `nukeVpc` => https://github.com/gruntwork-io/cloud-nuke/blob/migration/vpc-ec2/aws/resources/ec2_vpc.go#L774

That's why I am wondering what's the overall strategy of having this logic in multiple places?

Reply:
If the linter finds a `func` that is not being called, it is marked as greyed out/unused. I suspect this is leftover code from a previous refactor. IMHO, you can safely remove it if it is not being called.

Reply:
Thanks for refactoring the code. Can you have a separate PR to address the refactoring code and make this just focus on migrating to sdk-v2 so we can easily tell the issue if something comes up in the future? 

Just CC me for review. I can quickly go over the change. Runnign the unit tests should be able to verify if those are safe to get rid of.

Reply:
Hey @james03160927. I've spun off the following two PRs:

- https://github.com/gruntwork-io/cloud-nuke/pull/801
- https://github.com/gruntwork-io/cloud-nuke/pull/803

The https://github.com/gruntwork-io/cloud-nuke/pull/801 PR cherry-picks the commits from the PR that related to the TGW route tables. The https://github.com/gruntwork-io/cloud-nuke/pull/803 PR removes all of the unneeded EC2 code.

I will rebase this PR once those two are merged.